### PR TITLE
Enable warnings as errors in compiler component on OSX

### PR DESCRIPTION
**Pull request by fjeremic**
The commits have addressed all warnings on an OSX build, so to prevent
new warnings from being introduced we enable warnings as errors only on
this platform. Future platforms will follow. See each commit for an
explanation of what warning was triggered and fixed.

Issue: #1640

### CODE REVIEW

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build osx

**Comment by fjeremic** [comment]
Sadly this is going to be a bit more painful than I thought. My local clang version passes with no errors, but the version we use on the build farm does not. I can't revert to an older version because of my OS level. Going to be working on the farm machine to nail down the last set of errors.

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
@dsouzai do you want to be a committer for this one?

**Review by dsouzai — COMMENTED** [review]
LGTM; just awaiting clarification on the `-Wcomment` question.

File: compiler/optimizer/LoopVersioner.cpp [hunk 1]
```diff
@@ -9170,7 +9170,7 @@ bool TR_LoopVersioner::guardOkForExpr(TR::Node *node, bool onlySearching)
  * While it is not necessarily incorrect for \p node for to be unrepresentable,
  * it is unexpected and likely to cause a performance problem. So in case
  * \p node \em is unrepresentable, this method increments a static debug
- * counter matching <tt>{loopVersioner.unrepresentable/*}</tt>, and optionally
```
[the remark this answers is not in the document]
Comment (fjeremic) [reply compiler/optimizer/LoopVersioner.cpp:9173]: I also looked if there was a way to circumvent the `-Wcomment` warning and there is not, unless we add a bunch of `#pragma` code on for the various compilers which may be too overkill.

[the remark this answers is not in the document]
Comment (dsouzai) [reply compiler/optimizer/LoopVersioner.cpp:9173]: Yeah, I agree; Figured it was worth at least mentioning for posterity.

**Pull request merged**